src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -49940,11 +49963,186 @@ static void F()
+                Diagnostic(ErrorCode.ERR_StructLayoutCycle, "F").WithArguments("S.F", "S").WithLocation(4, 16),
+                // (12,9): warning CS8602: Dereference of a possibly null reference.
+                //         x.P.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x.P").WithLocation(12, 9));
+        }
+
+        [Fact]
+        [WorkItem(29619, "https://github.com/dotnet/roslyn/issues/29619")]
+        public void StructField_Cycle_02()
+        {
+            var source0 =
+@".class public sealed S extends [mscorlib]System.ValueType
+{
+  .field public valuetype [mscorlib]System.Nullable`1<valuetype S> F
+}";
+            var ref0 = CompileIL(source0);
+            var source =
+@"class Program
+{
+    static void F(S x, S y)
+    {
+        if (y.F == null) return;
+        _ = x.F.Value; // 1
+        _ = y.F.Value;
+    }
+}";
+            var comp = CreateCompilation(new[] { source }, new[] { ref0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (6,13): warning CS8629: Nullable value type may be null.
+                //         _ = x.F.Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "x.F").WithLocation(6, 13));
+        }
+
+        [Fact]
+        [WorkItem(29619, "https://github.com/dotnet/roslyn/issues/29619")]
+        public void StructField_Cycle_03()
+        {
+            var source0 =
+@".class public sealed S extends [mscorlib]System.ValueType
+{
+  .field public valuetype S F
+  .field public object G
+}";
+            var ref0 = CompileIL(source0);
+            var source =
+@"class Program
+{
+    static void F(S s)
+    {
+        s.G = null;
+        s.G.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(new[] { source }, new[] { ref0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (6,9): warning CS8602: Dereference of a possibly null reference.
+                //         s.G.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s.G").WithLocation(6, 9));
+        }
+
+        [Fact]
+        [WorkItem(29619, "https://github.com/dotnet/roslyn/issues/29619")]
+        public void StructField_Cycle_04()
+        {
+            var source0 =
+@".class public sealed S extends [mscorlib]System.ValueType
+{
+  .field public valuetype S F
+  .method public instance object get_P() { ldnull ret }
+  .method public instance void set_P(object 'value') { ret }
+  .property instance object P()
+  {
+    .get instance object S::get_P()
+    .set instance void S::set_P(object)
+  }
+}";
+            var ref0 = CompileIL(source0);
+            var source =
+@"class Program
+{
+    static void F(S s)
+    {
+        s.P = null;
+        s.P.ToString(); // 1
+    }
+}";
+            var comp = CreateCompilation(new[] { source }, new[] { ref0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
```

Comment:
>comp.VerifyDiagnostics(); [](start = 12, length = 25)

Do we expect any diagnostics here? It looks like there is a difference in behavior by comparison to source only scenario. Do we understand why?

Reply:
This is the expected behavior. The difference is `EmptyStructTypeCache.IsEmptyStructType()` returns different values for the source and metadata cases.

---
In reply to: [285218545](https://github.com/dotnet/roslyn/pull/35765#discussion_r285218545) [](ancestors = 285218545)

Reply:
> The difference is EmptyStructTypeCache.IsEmptyStructType() returns different values for the source and metadata cases.

Is the cycle broken for source in some different way?

---
In reply to: [285219414](https://github.com/dotnet/roslyn/pull/35765#discussion_r285219414) [](ancestors = 285219414,285218545)

Reply:
From source, cyclic fields are dropped from `EmptyStructTypeCache.GetStructInstanceFields()`.

---
In reply to: [285220520](https://github.com/dotnet/roslyn/pull/35765#discussion_r285220520) [](ancestors = 285220520,285219414,285218545)